mtags/src/main/scala-3/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -0,0 +1,416 @@
+        val autoImport = autoImportsGen.forSymbol(sym)
+        autoImport match
+          case Some(value) =>
+            visit(sym.info.dealias.typeSymbol, sym.decodedName, value)
+          case scala.None =>
+            visit(sym.info.dealias.typeSymbol, sym.showFullName, Nil)
+      else ()
+    }
+
+    // Step 3: special handle case when selector is a tuple or `FunctionN`.
+    if definitions.isTupleClass(selectorSym) || definitions.isFunctionClass(
+        selectorSym
+      )
+    then
+      val label =
+        if !patternOnly then s"case ${parents.selector.show} =>"
+        else parents.selector.show
+      result += CompletionValue.CaseKeyword(
+        selectorSym,
+        label,
+        Some(
+          if !patternOnly then
+            if config.isCompletionSnippetsEnabled() then "case ($0) =>"
+            else "case () =>"
+          else if config.isCompletionSnippetsEnabled() then "($0)"
+          else "()"
+        ),
+        Nil,
+        range = Some(completionPos.toEditRange),
+        command = config.parameterHintsCommand().asScala,
+      )
+    end if
+
+    val res = result.result()
+    res
+  end contribute
+
+  /**
+   * A `match` keyword completion to generate an exhaustive pattern match for sealed types.
+   *
+   * @param selector the match expression being deconstructed or `EmptyTree` when
+   *                 not in a match expression (for example `List(1).foreach { case@@ }`.
+   * @param completionPos the position of the completion
+   * @param typedtree typed tree of the file, used for generating auto imports
+   * @param indexedContext
+   * @param config
+   */
+  def matchContribute(
+      selector: Tree,
+      completionPos: CompletionPos,
+      indexedContext: IndexedContext,
+      config: PresentationCompilerConfig,
+  ): List[CompletionValue] =
+    import indexedContext.ctx
+    val pos = completionPos.sourcePos
+    val typedTree = indexedContext.ctx.compilationUnit.tpdTree
+    val definitions = indexedContext.ctx.definitions
+    val text = pos.source.content().mkString
+    val clientSupportsSnippets = config.isCompletionSnippetsEnabled()
+    lazy val autoImportsGen = AutoImports.generator(
+      pos,
+      text,
+      typedTree,
+      indexedContext,
+      config,
+    )
+    val completionGenerator = CompletionValueGenerator(
+      indexedContext,
+      completionPos,
+      clientSupportsSnippets,
+    )
+    val result = ListBuffer.empty[CompletionValue]
+    val tpe = selector.tpe.widen.bounds.hi match
+      case tr @ TypeRef(_, _) => tr.underlying
+      case t => t
+
+    def subclassesForType(tpe: Type)(using Context): List[Symbol] =
+      val subclasses = ListBuffer.empty[Symbol]
+      val parents = ListBuffer.empty[Symbol]
+      // Get parent types from refined type
+      def loop(tpe: Type): Unit =
+        tpe match
+          case AndType(tp1, tp2) => // for refined type like A with B with C
```

Comment:
Why this is needed? Is there any test case that related to this?

Reply:
> First and foremost, I don't think we don't need to provide an exhaustive match for non-sealed classes like this. 
We are not, it's only for single `case` completions.

Reply:
> Does this work if the class definitions are different files? (I don't think so)

It works if the class is imported or in the same package, so is visible in scope

Reply:
> I think children would provide the direct subclasses (maybe?)

Yes, but only for sealed types. For exhaustive match we can use `children` instead of `sealedStrictDescendants`, but when a child is a sealed trait should we split it even more?

Reply:
> Why this is needed? Is there any test case that related to this?

Yes, this one https://github.com/scalameta/metals/blob/15b8787c0f08520ef3bf54071b6ed8d318d9a703/tests/cross/src/test/scala/tests/pc/CompletionMatchSuite.scala#L274-L277